skimage/filters/thresholding.py

```diff
@@ -713,3 +716,283 @@ def threshold_triangle(image, nbins=256):
         arg_level = nbins - arg_level - 1
 
     return bin_centers[arg_level]
+
+
+def _mean_std(image, w):
+    """Return local mean and standard deviation of each pixel using a
+    neighborhood defined by a rectangular window with size w times w.
+    The algorithm uses integral images to speedup computation. This is
+    used by threshold_niblack and threshold_sauvola.
+
+    Parameters
+    ----------
+    image : (N, M) ndarray
+        Input image.
+    w : int
+        Odd window size (e.g. 3, 5, 7, ..., 21, ...).
+
+    Returns
+    -------
+    m : 2-D array of same size of image containning local mean values.
+    s : 2-D array of same size of image containning local standard
+        deviation values.
+
+    References
+    ----------
+    .. [1] F. Shafait, D. Keysers, and T. M. Breuel, "Efficient
+           implementation of local adaptive thresholding techniques
+           using integral images." in Document Recognition and
+           Retrieval XV, (San Jose, USA), Jan. 2008.
+    """
+    from skimage.transform.integral import integral_image
+
+    if w == 1 or w % 2 == 0:
+        raise ValueError(
+            "Window size w = %s must be odd and greater than 1." % (w))
+    I = integral_image(image)  # Integral Image.
+
+    # Pad left and top of Integral image with zeros
+    I = np.vstack((np.zeros((1, I.shape[1]), I.dtype), I))
+    I = np.hstack((np.zeros((I.shape[0], 1), I.dtype), I))
+
+    kern = np.zeros((w + 1, w + 1))
+    kern[0, 0], kern[-1, -1] = 1, 1
+    kern[[0, -1], [-1, 0]] = -1
+    # w2 holds total of pixels in window for each pixel (usually w x w).
+    w2 = ndi.convolve(
+        np.ones(image.shape, np.float), np.ones((w, w)), mode='constant')
+
+    m = ndi.convolve(I, kern, mode='nearest')[:-1, :-1] / w2
+    g = image.astype(np.float)
+    g2 = g ** 2.
+    m2 = m ** 2.
+    sum_g2 = ndi.convolve(g2, np.ones((w, w)), mode='constant')
+    sum_m2 = w2 * m2
+    s2 = (sum_g2 - sum_m2) / w2
+    s = np.sqrt(s2)
+    return m, s
```

Comment:
this function could be made 3-D compatible at small cost (``integral_image`` is 3-D compatible), hence ``threshold_niblack`` and `threshold_savola`` as well.